File: src/main/java/de/dc/planetsudo/game/strategy/DivineStrategy.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package de.dc.planetsudo.game.strategy;

import de.dc.planetsudo.level.levelobjects.AgentInterface;
import de.dc.planetsudo.level.levelobjects.Resource.ResourceType;

/**
 *
 * @author divine
 */
public class DivineStrategy extends AbstractStrategy {

	public DivineStrategy() {
	}
	
	public DivineStrategy(AgentInterface a) {
		super(a);
	}

	/**
	 * Wie viele Agenten sollen erstellt werde wird hier angegeben.
	 * @return Anzahl der Agenten
	 */
	@Override
	protected int loadAgentCount() {
		return 5;
	}

	@Override
	protected void loadRules() {
		//-------------------------------------------->
		createRule(new Rule(0, "Just Go") {
			@ Override
			protected boolean constraint() {
				return true;
			}
			@ Override
			protected void action() {
				agent.go();
				
			}
		});
		//-------------------------------------------->
		createRule(new Rule(1, "Discover") {
			@ Override
			protected boolean constraint() {
				return agent.isCommander();
			}
			@ Override
			protected void action() {
				agent.goRight(3);
			}
		});
		//-------------------------------------------->
		createRule(new Rule(5, "Go to Marker") {
			@ Override
			protected boolean constraint() {
				return !agent.isCommander() && mothership.isMarkerDeployed();
			}
			@ Override
			protected void action() {
				agent.goToMarker();
			}
		});
		//-------------------------------------------->
		createRule(new Rule(10, "Search") {
			@ Override
			protected boolean constraint() {
				return agent.seeMarker();
			}
			@ Override
			protected void action() {
				mothership.clearMarker();
				agent.searchResources();
			}
		});
		//-------------------------------------------->
		createRule(new Rule(20, "See Resources") {
			@ Override
			protected boolean constraint() {
				return !agent.isCommander() &&agent.seeResource();
			}
			@ Override
			protected void action() {
				agent.goToResource();
			}
		});
		//-------------------------------------------->
		createRule(new Rule(30, "PickUp 1P Resource") {
			@ Override
			protected boolean constraint() {
				return !agent.isCommander() && agent.isTouchingResource(ResourceType.Normal);
			}
			@ Override
			protected void action() {
				agent.pickupResource();
			}
		});
		//-------------------------------------------->
		createRule(new Rule(31, "PickUp and Place") {
			@ Override
			protected boolean constraint() {
				return !agent.isCommander() && (agent.isTouchingResource(ResourceType.ExtremPoint) || agent.isTouchingResource(ResourceType.DoublePoints) || agent.isTouchingResource(ResourceType.ExtraMothershipFuel));
			}
			@ Override
			protected void action() {
				if(!mothership.isMarkerDeployed()) {
					agent.deployMarker();
				}
				agent.pickupResource();
			}
		});
		//-------------------------------------------->
		createRule(new Rule(32, "PickUp 5P and Place") {
			@ Override
			protected boolean constraint() {
				return agent.isTouchingResource(ResourceType.ExtremPoint);
			}
			@ Override
			protected void action() {
				agent.deployMarker();
				agent.pickupResource();
			}
		});
		//-------------------------------------------->
		createRule(new Rule(33, "Ignore") {
			@ Override
			protected boolean constraint() {
				return agent.isTouchingResource(ResourceType.Mine) || agent.isTouchingResource(ResourceType.ExtraAgentFuel);
			}
			@ Override
			protected void action() {
				agent.go();
			}
		});


		//-------------------------------------------->
		createRule(new Rule(34, "Secure") {
			@ Override
			protected boolean constraint() {
				return agent.isGameOverSoon();
			}
			@ Override
			protected void action() {
				agent.goToMothership();
			}
		});
		//-------------------------------------------->
		createRule(new Rule(35, "Saved") {
			@ Override
			protected boolean constraint() {
				return agent.isGameOverSoon() && agent.isAtMothership();
			}
			@ Override
			protected void action() {

			}
		});
		//-------------------------------------------->
		createRule(new Rule(36, "Order Fuel") {
			@ Override
			protected boolean constraint() {
				return agent.isAtMothership() && agent.isGameOverSoon() && mothership.hasFuel()&& agent.getFuelInPercent() < 100;
			}
			@ Override
			protected void action() {
				agent.orderFuel(100);
			}
		});
		//-------------------------------------------->
		createRule(new Rule(39, "Support Agent") {
			@ Override
			protected boolean constraint() {
				return mothership.needSomeoneSupport() && !agent.isSupportOrdered();
			}
			@ Override
			protected void action() {
				agent.goToSupportAgent();
			}
		});
		//-------------------------------------------->
		createRule(new Rule(40, "Save Resource") {
			@ Override
			protected boolean constraint() {
				return agent.isCarringResource();
			}
			@ Override
			protected void action() {
				agent.goToMothership();
			}
		});
		//-------------------------------------------->
		createRule(new Rule(60, "HelpLostAgent") {
			@ Override
			protected boolean constraint() {
				return agent.seeLostTeamAgent();
			}
			@ Override
			protected void action() {
				agent.spendTeamAgentFuel(300);
			}
		});
		//-------------------------------------------->
		createRule(new Rule(90, "FightAgainstMothership") {
			@ Override
			protected boolean constraint() {
				return agent.seeAdversaryMothership();
			}
			@ Override
			protected void action() {
				agent.fightWithAdversaryMothership();
			}
		});
		//-------------------------------------------->
		createRule(new Rule(91, "FightAgainstMothership & Order Support") {
			@ Override
			protected boolean constraint() {
				return !agent.isSupportOrdered() && agent.seeAdversaryMothership();
			}
			@ Override
			protected void action() {
				agent.fightWithAdversaryMothership();
				agent.orderSupport();
			}
		});
		//-------------------------------------------->
		createRule(new Rule(100, "TurnToAdversaryAgent") {
			@ Override
			protected boolean constraint() {
				return agent.isUnderAttack();
			}
			@ Override
			protected void action() {
				agent.turnLeft(60);
			}
		});
		//-------------------------------------------->
		createRule(new Rule(110, "SaveMothership") {
			@ Override
			protected boolean constraint() {
				return mothership.isDamaged();
			}
			@ Override
			protected void action() {
				agent.goToMothership();
			}
		});
		//-------------------------------------------->
		createRule(new Rule(120, "RepaireMothership") {
			@ Override
			protected boolean constraint() {
				return mothership.isDamaged() && agent.isAtMothership();
			}
			@ Override
			protected void action() {
				agent.repairMothership();
			}
		});
		//-------------------------------------------->
		createRule(new Rule(190, "FightAgainstAgent") {
			@ Override
			protected boolean constraint() {
				return agent.seeAdversaryAgent();
			}
			@ Override
			protected void action() {
				agent.fightWithAdversaryAgent();
			}
		});
		//-------------------------------------------->
		createRule(new Rule(195, "PlaceMine") {
			@ Override
			protected boolean constraint() {
				return agent.hasMine() && agent.isUnderAttack();
			}
			@ Override
			protected void action() {
				agent.deployMine();
				agent.goLeft(180);
			}
		});
		//-------------------------------------------->
		createRule(new Rule(200, "GoBackToMothership") {
			@ Override
			protected boolean constraint() {
				return agent.getFuel() < 300;
			}
			@ Override
			protected void action() {
				agent.goToMothership();
			}
		});
		//-------------------------------------------->
		createRule(new Rule(400, "OrderFuel") {
			@ Override
			protected boolean constraint() {
				return mothership.hasFuel() && (agent.getFuelInPercent() < 90) && (agent.isAtMothership());
			}
			@ Override
			protected void action() {
				agent.orderFuel(100);
			}
		});
		//-------------------------------------------->
		createRule(new Rule(500, "OrderFuelDuringFight") {
			@ Override
			protected boolean constraint() {
				return mothership.hasFuel() && (agent.getFuel() < 100) && agent.isUnderAttack() && agent.isAtMothership();
			}
			@ Override
			protected void action() {
				agent.orderFuel(5);
				agent.goLeft(10);
			}
		});
		//-------------------------------------------->
		createRule(new Rule(501, "OrderFuelDuringFight") {
			@ Override
			protected boolean constraint() {
				return mothership.hasFuel() && (agent.getFuel() < 100) && agent.seeAdversaryAgent() && agent.isAtMothership();
			}
			@ Override
			protected void action() {
				agent.orderFuel(5);
				agent.fightWithAdversaryAgent();
			}
		});
		//-------------------------------------------->
		createRule(new Rule(550, "Pass Resource") {
			@ Override
			protected boolean constraint() {
				return agent.isCarringResource() && agent.isAtMothership();
			}
			@ Override
			protected void action() {
				agent.deliverResourceToMothership();
			}
		});
		//-------------------------------------------->
		createRule(new Rule(600, "CallForHelp") {
			@ Override
			protected boolean constraint() {
				return (!agent.isSupportOrdered()) && ((agent.getFuel() < 5) || agent.isUnderAttack());
			}
			@ Override
			protected void action() {
				agent.orderSupport();
			}
		});
		//-------------------------------------------->
		createRule(new Rule(800, "PickUp Fuel") {
			@ Override
			protected boolean constraint() {
				return agent.getFuelInPercent() < 50 && agent.isTouchingResource(ResourceType.ExtraAgentFuel);
			}
			@ Override
			protected void action() {
				agent.pickupResource();
			}
		});
		//-------------------------------------------->
		createRule(new Rule(900, "Follow Wall") {
			@ Override
			protected boolean constraint() {
				return agent.isCommander() && agent.isCollisionDetected();
			}
			@ Override
			protected void action() {
				agent.turnLeft(3);
				
			}
		});
		//-------------------------------------------->
		createRule(new Rule(1000, "AvoidWall") {
			@ Override
			protected boolean constraint() {
				return !agent.isCommander() && agent.isCollisionDetected();
			}
			@ Override
			protected void action() {
				agent.turnRandom(150);

			}
		});
//		//-------------------------------------------->
//		createRule(new Rule(1001, "Cancel Support") {
//			@ Override
//			protected boolean constraint() {
//				return agent.isSupportOrdered() && !agent.seeAdversaryMothership() && agent.getFuel() > 10 && !agent.isUnderAttack() && !agent.seeAdversaryAgent();
//			}
//			@ Override
//			protected void action() {
//				agent.cancelSupport();
//
//			}
//		});
//		//-------------------------------------------->
	}
}

File: src/main/kotlin/org/openbase/planetsudo/game/strategy/DivineStrategy.kt
package org.openbase.planetsudo.game.strategy

import org.openbase.planetsudo.game.SwatTeam
import org.openbase.planetsudo.level.levelobjects.AgentInterface
import org.openbase.planetsudo.level.levelobjects.Resource.ResourceType
import org.openbase.planetsudo.level.levelobjects.Tower

/**
 *
 * @author [Divine Threepwood](mailto:divine@openbase.org)
 */
class DivineStrategy(agent: AgentInterface) : StrategyLevelLegacy(agent) {

    /**
     * Hier wird angegeben wie viele Agenten dem Team zur Verfügung stehen sollen.
     *
     * @return Anzahl der Agenten
     */
    override fun loadAgentCount(): Int {
        return 7
    }

    override fun loadSwatTeams() {
        createSwat(SwatTeam.ALPHA, 1)
    }

    override fun loadRules() {
        // -------------------------------------------->
        createRule(
            object : Rule("Just Go", SwatTeam.ALL) {
                override fun constraint(): Boolean {
                    return true
                }

                override fun action() {
                    agent.go()
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("Discover", SwatTeam.COMMANDER) {
                override fun constraint(): Boolean {
                    return true
                }

                override fun action() {
                    agent.goRight(4)
                }
            },
        )
        // -------------------------------------------->
//        createRule(
//            object : Rule("Avoid Agents", SwatTeam.COMMANDER) {
//                override fun constraint(): Boolean {
//                    return agent.seeTeamAgent
//                }
//
//                override fun action() {
//                    agent.turnRandom()
//                }
//            },
//        )
        // -------------------------------------------->
        "See Resources" swat SwatTeam.NOT_COMMANDER inCase {
            agent.seeResource &&
                (agent.isTonicAtLimit && agent.seeResource(ResourceType.Tonic)).not()
        } then {
            agent.goToResource()
        }
        // -------------------------------------------->
        createRule(
            object : Rule("PickUp 1P Resource") {
                override fun constraint(): Boolean {
                    return !agent.isCommander && agent.isTouchingResource(ResourceType.Normal)
                }

                override fun action() {
                    agent.pickupResource()
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("Go to Marker") {
                override fun constraint(): Boolean {
                    return !agent.isCommander && mothership.isMarkerDeployed
                }

                override fun action() {
                    agent.goToMarker()
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("Search") {
                override fun constraint(): Boolean {
                    return !agent.isCommander && agent.seeMarker()
                }

                override fun action() {
                    mothership.clearMarker()
                    agent.searchResources()
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("PickUp") {
                override fun constraint(): Boolean {
                    return !agent.isCommander && (
                        agent.isTouchingResource(ResourceType.DoublePoints) || agent.isTouchingResource(
                            ResourceType.ExtraMothershipFuel,
                        )
                        )
                }

                override fun action() {
                    agent.pickupResource()
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("PickUp and Place") {
                override fun constraint(): Boolean {
                    return !agent.isCommander && agent.isTouchingResource(ResourceType.ExtremPoint)
                }

                override fun action() {
                    if (!mothership.isMarkerDeployed) {
                        agent.deployMarker()
                    }
                    agent.pickupResource()
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("PickUp and Place") {
                override fun constraint(): Boolean {
                    return agent.isCommander && (
                        agent.isTouchingResource(ResourceType.DoublePoints) || agent.isTouchingResource(
                            ResourceType.ExtraMothershipFuel,
                        )
                        ) && !mothership.isMarkerDeployed && !agent.seeMarker()
                }

                override fun action() {
                    agent.deployMarker()
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("PickUp 5P and Place") {
                override fun constraint(): Boolean {
                    return !agent.isCommander && agent.isTouchingResource(ResourceType.ExtremPoint)
                }

                override fun action() {
                    agent.deployMarker()
                    agent.pickupResource()
                }
            },
        )
        // -------------------------------------------->
        "PickUp Tonic" all inCase { agent.isTouchingResource(ResourceType.Tonic) && !agent.isTonicAtLimit } then {
            agent.pickupResource()
        }
        // -------------------------------------------->
        createRule(
            object : Rule("Discover 5P and Place Marker", SwatTeam.COMMANDER) {
                override fun constraint(): Boolean {
                    return agent.isTouchingResource(ResourceType.ExtremPoint) && !agent.seeMarker()
                }

                override fun action() {
                    agent.deployMarker()
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("Place Tower", SwatTeam.COMMANDER) {
                override fun constraint(): Boolean {
                    return agent.isTouchingResource(ResourceType.ExtremPoint) && agent.hasTower
                }

                override fun action() {
                    agent.constructTower(Tower.TowerType.ObservationTower)
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("Ignore") {
                override fun constraint(): Boolean {
                    return agent.isTouchingResource(ResourceType.Mine) || agent.isTouchingResource(ResourceType.ExtraAgentFuel)
                }

                override fun action() {
                    agent.go()
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("Secure") {
                override fun constraint(): Boolean {
                    return agent.isGameOverSoon && !agent.isAtMothership
                }

                override fun action() {
                    agent.shift()
                    agent.goToMothership()
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("Mark Resource") {
                override fun constraint(): Boolean {
                    return agent.isCarryingResource && !mothership.isMarkerDeployed && agent.seeResource
                }

                override fun action() {
                    agent.deployMarker()
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("Saved") {
                override fun constraint(): Boolean {
                    return agent.isGameOverSoon && agent.isAtMothership
                }

                override fun action() {
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("Order Fuel") {
                override fun constraint(): Boolean {
                    return agent.isAtMothership && !agent.isGameOverSoon && mothership.hasFuel() && agent.fuelInPercent < 90
                }

                override fun action() {
                    agent.orderFuel(100)
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("Support Agent") {
                override fun constraint(): Boolean {
                    return mothership.needSomeoneSupport() && !agent.hasRequestedSupport
                }

                override fun action() {
                    agent.goToSupportAgent()
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("Save Resource") {
                override fun constraint(): Boolean {
                    return agent.isCarryingResource && !agent.isAtMothership
                }

                override fun action() {
                    agent.goToMothership()
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("Shift for Extra Resource") {
                override fun constraint(): Boolean {
                    return agent.isCarryingResource(ResourceType.ExtremPoint) && agent.hasTonic && !agent.isShifting && !agent.isAtMothership
                }

                override fun action() {
                    agent.shift()
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("HelpLostAgent") {
                override fun constraint(): Boolean {
                    return agent.seeLostTeamAgent() && !teamAgent.isAtMothership
                }

                override fun action() {
                    agent.spendTeamAgentFuel(300)
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("FightAgainstMothership") {
                override fun constraint(): Boolean {
                    return agent.seeEnemyMothership
                }

                override fun action() {
                    agent.fightWithEnemyMothership()
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("FightAgainstMothership & Order Support") {
                override fun constraint(): Boolean {
                    return !agent.hasRequestedSupport && agent.seeEnemyMothership
                }

                override fun action() {
                    agent.fightWithEnemyMothership()
                    agent.requestSupport()
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("TurnToAdversaryAgent") {
                override fun constraint(): Boolean {
                    return agent.isUnderAttack
                }

                override fun action() {
                    agent.turnLeft(60)
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("SaveMothership") {
                override fun constraint(): Boolean {
                    return mothership.isDamaged && !agent.isAtMothership
                }

                override fun action() {
                    agent.goToMothership()
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("RepaireMothership") {
                override fun constraint(): Boolean {
                    return mothership.isDamaged && agent.isAtMothership
                }

                override fun action() {
                    agent.repairMothership()
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("FightAgainstAgent") {
                override fun constraint(): Boolean {
                    return agent.seeEnemyAgent
                }

                override fun action() {
                    agent.fightWithEnemyAgent()
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("PlaceMine") {
                override fun constraint(): Boolean {
                    return agent.hasMine && agent.isUnderAttack
                }

                override fun action() {
                    agent.deployMine()
                    agent.goLeft(180)
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("GoBackToMothership") {
                override fun constraint(): Boolean {
                    return agent.fuel < 300 && !agent.isAtMothership
                }

                override fun action() {
                    agent.goToMothership()
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("OrderFuel") {
                override fun constraint(): Boolean {
                    return mothership.hasFuel() && (agent.fuelInPercent < 90) && (agent.isAtMothership)
                }

                override fun action() {
                    agent.orderFuel(100)
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("OrderFuelDuringFight") {
                override fun constraint(): Boolean {
                    return mothership.hasFuel() && (agent.fuel < 100) && agent.isUnderAttack && agent.isAtMothership
                }

                override fun action() {
                    agent.orderFuel(5)
                    agent.goLeft(10)
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("OrderFuelDuringFight") {
                override fun constraint(): Boolean {
                    return mothership.hasFuel() && (agent.fuel < 100) && agent.seeEnemyAgent && agent.isAtMothership
                }

                override fun action() {
                    agent.orderFuel(5)
                    agent.fightWithEnemyAgent()
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("Pass Resource") {
                override fun constraint(): Boolean {
                    return agent.isCarryingResource && agent.isAtMothership
                }

                override fun action() {
                    agent.transferResourceToMothership()
                    agent.turnAround()
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("CallForHelpDuringFight") {
                override fun constraint(): Boolean {
                    return !agent.hasRequestedSupport && agent.isUnderAttack
                }

                override fun action() {
                    agent.requestSupport()
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("CallForHelp") {
                override fun constraint(): Boolean {
                    return !agent.hasRequestedSupport && !agent.isAtMothership && agent.fuel < 5
                }

                override fun action() {
                    agent.requestSupport()
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("PickUp Fuel") {
                override fun constraint(): Boolean {
                    return agent.fuelInPercent < 50 && agent.isTouchingResource(ResourceType.ExtraAgentFuel)
                }

                override fun action() {
                    agent.pickupResource()
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("Follow Wall", SwatTeam.COMMANDER) {
                override fun constraint(): Boolean {
                    return agent.isCollisionDetected
                }

                override fun action() {
                    agent.turnLeft(4)
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("Construct Tower", SwatTeam.COMMANDER) {
                override fun constraint(): Boolean {
                    return agent.hasTower && agent.seeResource
                }

                override fun action() {
                    agent.constructTower(Tower.TowerType.DefenceTower)
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("Follow Wall", SwatTeam.ALPHA) {
                override fun constraint(): Boolean {
                    return agent.isCollisionDetected
                }

                override fun action() {
                    agent.turnRight(4)
                }
            },
        )
        // -------------------------------------------->
        "Spy" all inCase { agent.isInvisible && agent.seeEnemyAgent } then {
            agent.goToEnemyAgent()
        }
        "Attack Mothership" all inCase { agent.isInvisible && agent.seeEnemyMothership } then {
            agent.fightWithEnemyMothership()
        }
        "Attack Agent" all inCase { agent.isInvisible && agent.seeEnemyAgent && agent.seeEnemyMothership } then {
            agent.fightWithEnemyAgent()
        }
        "Prepare for Attack" all inCase { agent.isInvisible && agent.seeEnemyMothership && agent.actionPoints < 10000 } then {
            // wait for APs
        }
        "Call Force" all inCase { agent.isInvisible && agent.seeEnemyMothership && !agent.hasRequestedSupport } then {
            agent.requestSupport()
        }
        // -------------------------------------------->
        "Make invisible" all inCase { agent.tonicInPercent == 100 && agent.isVisible && agent.isFighting } then {
            agent.makeInvisible()
        }
        // -------------------------------------------->
        createRule(
            object : Rule("AvoidWall", SwatTeam.NOT_ALPHA, SwatTeam.NOT_COMMANDER) {
                override fun constraint(): Boolean {
                    return agent.isCollisionDetected
                }

                override fun action() {
                    agent.turnRandom(150)
                }
            },
        )
        // -------------------------------------------->
        createRule(
            object : Rule("Cancel Support") {
                override fun constraint(): Boolean {
                    return agent.hasRequestedSupport && !agent.seeEnemyMothership && agent.fuel > 10 && !agent.isUnderAttack && !agent.seeEnemyAgent
                }

                override fun action() {
                    agent.cancelSupport()
                }
            },
        )
        // -------------------------------------------->
    }
}
